frontend/src/app/user/login/login.component.ts

```diff
@@ -1,82 +1,73 @@
+    public usernameErrors$ = controlErrorMessages$(this.form, 'username');
+    public passwordErrors$ = controlErrorMessages$(this.form, 'password');
+    public formErrors$ = formErrorMessages$(this.form);
 
-    private returnUrl: string;
+    public loading$ = merge(
+        this.authService.login$.pipe(map(() => true)),
+        this.authService.loginResult$.pipe(map(() => false)),
+    ).pipe(startWith(false));
 
     constructor(
         private authService: AuthService,
-        private activatedRoute: ActivatedRoute,
+        private toastService: ToastService,
+        private destroyRef: DestroyRef,
         private router: Router,
-    ) {
-        this.returnUrl = this.activatedRoute.snapshot.queryParams['returnUrl'] || '/';
-    }
+    ) { }
 
-    submit() {
-        const valid = this.checkValidation();
-        if (valid) {
-            this.authService.login(
-                this.usernameInput.value as string, this.passwordInput.value as string
-            ).subscribe({
-                next: () => this.loginSucces(),
-                error: (e: HttpErrorResponse) => this.loginFailed(e)
-            });
-        }
+    ngOnInit(): void {
+        this.authService.loginResult$
+            .pipe(takeUntilDestroyed(this.destroyRef))
+            .subscribe(result => {
+                if ('error' in result) {
+                    setErrors(result.error, this.form);
+                } else {
+                    this.toastService.show({
+                        header: 'Sign in successful',
+                        body: 'You have been successfully signed in.',
+                        type: 'success'
+                    });
+                    this.router.navigate(['/']);
+                }
+            })
     }
 
-    private checkValidation() {
-        if (this.usernameInput.invalid) {
-            if (this.usernameInput.errors?.['required']) {
-                this.usernameErrorMessage = 'username is required';
-            } else {
-                this.usernameErrorMessage = 'invalid username';
-            }
-        } else {
-            this.usernameErrorMessage = '';
-        }
-
-        if (this.passwordInput.invalid) {
-            if (this.passwordInput.errors?.['required']) {
-                this.passwordErrorMessage = 'password is required';
-            } else {
-                this.passwordErrorMessage = 'invalid password';
-            }
-        } else {
-            this.passwordErrorMessage = '';
+    public submit(): void {
+        this.form.markAllAsTouched();
+        updateFormValidity(this.form);
+        if (!this.form.valid) {
+            return;
         }
-
-        return this.usernameInput.valid && this.passwordInput.valid;
-    }
-
-    private loginSucces() {
-        this.router.navigate([this.returnUrl]);
-    }
-
-    private loginFailed(error: HttpErrorResponse) {
-        this.requestFailed = true;
+        this.authService.login$.next(this.form.getRawValue());
```

Comment:
`.getRawValue()` instead of `.value` following the docs. When using `.value`, you only take values of non-disabled controls, and since all controls can be disabled programmatically, the type is `Partial<...>`. `getRawValue()` gets all values from all controls.